### PR TITLE
[GlobalIsel][NFC] Modernize UBFX combine

### PR DESCRIPTION
**Pull request by tschuett**
Credits: https://reviews.llvm.org/D99283

### CODE REVIEW

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 1]
```diff
@@ -4521,19 +4521,21 @@ bool CombinerHelper::matchBitfieldExtractFromSExtInReg(
...
   if (LI && !LI->isLegalOrCustom({TargetOpcode::G_UBFX, {Ty, ExtractTy}}))
     return false;
 
   int64_t AndImm, LSBImm;
   Register ShiftSrc;
   const unsigned Size = Ty.getScalarSizeInBits();
-  if (!mi_match(MI.getOperand(0).getReg(), MRI,
+  if (!mi_match(And->getReg(0), MRI,
                 m_GAnd(m_OneNonDBGUse(m_GLShr(m_Reg(ShiftSrc), m_ICst(LSBImm))),
                        m_ICst(AndImm))))
```
Comment (dc03-work) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:4540]: Can this condition not be removed now that you're matching it in the TableGen?

Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:4540]: 1. It is collecting registers and int64_t s.
2. `mi_match` returns a boolean. I don't want to and cannot ignore it.
3. I could add a comment: "Cannot fail due to pattern."

Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:4540]: "m_OneNonDBGUse" could cause failure.

Comment (dc03-work) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:4540]: > 1. It is collecting registers and int64_t s.
> 2. `mi_match` returns a boolean. I don't want to and cannot ignore it.
> 3. I could add a comment: "Cannot fail due to pattern."

1. You can add those as function parameters and pass them from the TableGen.
2. The only part of this match that differs from the TableGen is the `m_OneNonDBGUse` bit, which you can add as a separate condition. You can also use the `HasOneUse` predicate (#91578).

Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:4540]: I can only pass MachineOperands as parameters. The constants would still need a separate mechanism. The `HasOneUse` predicate is still limited to the instruction emitter and cannot be used in the combiner.

**Review by arsenm — APPROVED** [review]
This is improving but it's still unfortunate we have to repeat the pattern inside the function

File: llvm/test/CodeGen/AArch64/GlobalISel/form-bitfield-extract-from-and.mir [hunk 2]
```diff
@@ -6,6 +6,10 @@
 #
 # and (lshr x, cst), mask -> ubfx x, cst, width
 
+# LSB = 5
```
Comment (jayfoad) [inline llvm/test/CodeGen/AArch64/GlobalISel/form-bitfield-extract-from-and.mir:9]: Please precommit all the formatting changes in this files. I can't tell if there are any changes related to the patch.

Comment (tschuett) [reply llvm/test/CodeGen/AArch64/GlobalISel/form-bitfield-extract-from-and.mir:9]: The update script deleted the comments and I had to move them up to #.

Comment (jayfoad) [reply llvm/test/CodeGen/AArch64/GlobalISel/form-bitfield-extract-from-and.mir:9]: I understand that, but please commit all those formatting changes as a separate NFC patch.

**Pull request merged**